WebKit/WebKit
# Re-sync color conversion constants with latest in spec

#### 1d979e658a4948cb03f593c9091b70b275eff5c3
<pre>
Re-sync color conversion constants with latest in spec
<a href="https://bugs.webkit.org/show_bug.cgi?id=272952">https://bugs.webkit.org/show_bug.cgi?id=272952</a>

Reviewed by Darin Adler.

Update the following constants:
- xyzToLinear/linearToXYZ 3x3 matrices for the RGB color types.
- chromatic adaptation 3x3 matrices
- D50 and D65 white point vector
- OKLab XYZ &lt;-&gt; LMS matrices

Also updated polar/rectilinear conversions to correctly
handle the achromatic cases, ensuring conversion to lch/oklch
have &quot;powerless&quot; hue when chroma is close to 0, and when
that &quot;powerless&quot; hue is transformed back to lab/oklab, that
it the opponent A/B values are set to 0.

* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-relative-color.html:
Update one result to account to more closely match colorjs.io reference results.

* LayoutTests/imported/w3c/web-platform-tests/css/css-color/nested-color-mix-with-currentcolor-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-color-mix-function-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-relative-color-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-relative-color-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-mix-out-of-gamut-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/relative-color-out-of-gamut-expected.txt:
    - Updated results that are now passing.

* Tools/TestWebKitAPI/Tests/WebCore/ColorTests.cpp:
(TestWebKitAPI::TEST(Color, P3ConversionToSRGB)):
    - Updated result to account for change in OKLab conversion. (OKLab is used for the
      gamut mapping, which is why this only effected this one test).

* Source/WebCore/platform/graphics/ColorConversion.cpp:
(WebCore::convertToPolarForm):
    - Updated to account for sufficiently small a/b values that the result would be achromatic.

(WebCore::convertToRectangularForm):
    - Handle missing (NaN) hue, to correctly set a/b to 0.

(WebCore::OKLab&lt;float&gt;&gt;::convert):
    - Update XYZ &lt;-&gt; LMS constants to latest in spec.

* Source/WebCore/platform/graphics/ColorConversion.h:
    - Move and update white point constants to the header to be closer to the chromatic adaptation code.
    - Update chromatic adaptation constants and add a comment explaining what they are all about.

* Source/WebCore/platform/graphics/ColorMatrix.h:
(WebCore::ColorMatrix::ColorMatrix):
    - Allow conversions on construction to allow matrices to be defined with either double or float.

(WebCore::brightnessColorMatrix):
(WebCore::contrastColorMatrix):
(WebCore::invertColorMatrix):
(WebCore::opacityColorMatrix):
    - Fix a typo that was copied in a few places. compontent -&gt; component

* Source/WebCore/platform/graphics/ColorTypes.h:
    - Update LinearRGB &lt;-&gt; XYZ constants.

Canonical link: <a href="https://commits.webkit.org/278112@main">https://commits.webkit.org/278112@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/52ed0f3bba9dcb4b6f6b0096a0f592cbc790a75b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49532 "37 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28814 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/52576 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/52778 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/208 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51836 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34832 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/26434 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/40448 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/51632 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/26357 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/42673 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21563 "Found 1 new API test failure: /TestWebCore:Color.P3ConversionToSRGB (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23822 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43856 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/7908 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/45732 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/44361 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54350 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24619 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/20808 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/47819 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/25887 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/42833 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/46837 "Found 1 new API test failure: /TestWebCore:Color.P3ConversionToSRGB (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10876 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26735 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/25609 "Built successfully") | | |
